don't extend a second entity if we're not joining to the second entity

MegaphoneJon opened the pull request:
Some jerk (me) caused a regression with #332.

The Participant Extended Report adds a "related contact" to the report - and #332 fixed some incorrect results by not joining the related contact table if no related contact fields were selected.

However, ` CRM_Extendedreport_Form_Report_ExtendedReport::extendedCustomDataFrom()`, whose docblock reads, `Overridden to support custom data for multiple entities of the same type`, tries to join the custom field tables for the related contact regardless.  So if you try to include custom fields (on the primary contact), you get the error `Database Error Code: Unknown column 'related_civicrm_contact.id' in 'on clause', 1054`.

The solution is to check if the table the custom field extends is actually selected, and if not, skip adding it.

This revealed a follow-on bug - adding a related contact's email or phone fields without adding any contact fields would cause the same error.  So I fixed that in the ParticipantExtended report.

eileenmcnaughton commented:
@MegaphoneJon I've been started to add those checks into the from clauses - eg.

```
  protected function joinAddressFromContact($prefix = '', $extra = []) {

    if ($this->isTableSelected($prefix . 'civicrm_address')) {
```

Does it make more sense there?

MegaphoneJon commented:
Ooh, that looks more clever.  I'll investigate tomorrow.

MegaphoneJon commented:
OK, I rewrote this to include the conditional in the "from clause" function.  However, I ran into some trouble with this snippet in the `buildColumns()` method:

```php
    if (!$tableAlias) {
      $tableAlias = str_replace('civicrm_', '', $tableName);
    }
```

This causes table aliases to be inconsistent with what's in `CRM_Report_Form::_selectedTables`, and I suspect is purely for cosmetic (DX) reasons.  I think this was previously worked around by adding a `1 ||` to the `joinRelatedContactFromParticipant` (you can see I removed this) but this introduced the bug fixed in #332.

I believe that my fix handles all scenarios, but of course changing `buildColumns()` has a wider reach than these other functions.

eileenmcnaughton commented:
/test

eileenmcnaughton commented:
/test

eileenmcnaughton commented:
I think I'm mostly there with solving this - see https://github.com/eileenmcnaughton/nz.co.fuzion.extendedreport/pull/405 - although that builds on a bunch of other work
